segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/GarbageCollector.java

```diff
@@ -0,0 +1,182 @@
+
+    private final long delayInMillis;
+
+    private CompletableFuture<Void> loopFuture;
+
+    /**
+     * Constructs a new instance.
+     *
+     * @param chunkedSegmentStorage Instance of {@link ChunkedSegmentStorage}.
+     * @param config Configuration to use.
+     */
+    public GarbageCollector(ChunkedSegmentStorage chunkedSegmentStorage, ChunkedSegmentStorageConfig config) {
+        this.chunkedSegmentStorage = chunkedSegmentStorage;
+        this.config = config;
+        this.delayInMillis = Duration.ofSeconds(config.getGarbageCollectionFrequencyInSeconds()).toMillis();
+    }
+
+    /**
+     * Initializes this instance.
+     */
+    public void initialize() {
+        loopFuture = run();
+    }
+
+    public CompletableFuture<Void> run() {
+        return Futures.loop(
+                () -> !closed.get(),
+                () -> CompletableFuture.completedFuture(null)
+                        .thenComposeAsync(v -> deleteGarbage(true, config.getGarbageCollectionConcurrency()),
+                                chunkedSegmentStorage.getExecutor())
+                        .handleAsync((v, ex) -> {
+                            if (null != ex) {
+                                log.error("{} Error during run {}.", chunkedSegmentStorage.getLogPrefix(), ex);
+                            }
+                            return null;
+                        }), chunkedSegmentStorage.getExecutor());
+    }
+
+    /**
+     * Sets whether backgound claenup is suspended or not.
+     * @param value
+     */
+    void setSuspended(boolean value) {
+        suspended.set(value);
+    }
+
+    /**
+     * collect the garbage chunks.
+     *
+     * @param chunksToDelete List of chunks to delete.
+     */
+    CompletableFuture<Void> addToGarbage(Collection<String> chunksToDelete) {
+        val currentTime = System.currentTimeMillis();
+        chunksToDelete.forEach(chunkToDelete -> garbageChunks.put(chunkToDelete, currentTime));
+        return CompletableFuture.completedFuture(null);
+    }
+
+    /**
+     * Delete the garbage chunks.
+     *
+     * @param isBackground True if the caller is backgound task else False if called explicitly.
+     * @param maxItems Maximum number of items to delete at a time.
+     * @return CompletableFuture which is completed when garbage is deleted.
+     */
+    CompletableFuture<Void> deleteGarbage(boolean isBackground, int maxItems) {
+        if (suspended.get() && isBackground) {
+            return CompletableFuture.completedFuture(null);
+        }
+        val currentTime = System.currentTimeMillis();
+        val chunksToDelete = new ArrayList<String>();
+        int count = 0;
+        for (val entry : garbageChunks.entrySet()) {
+            if (canDelete(entry.getValue(), currentTime)) {
+                chunksToDelete.add(entry.getKey());
+                // Throttle how many chunks are deleted at once.
+                if (++count >= maxItems) {
```

Comment:
If you do a sorted list, you may even be able to compress this whole block into a `stream()` with limit and filter.

Reply:
out dated . But new code has different structure.